README.md

```diff
@@ -80,6 +80,10 @@ greater than zero, `go install` is invoked with `-mod=vendor`, causing the build
 to skip downloading and checking of dependencies. This results in only the
 dependencies from the top level `vendor` directory being used.
 
+If the file `bin/pre-compile` or `bin/post-compile` exists in the repository and
```

Comment:
I have been a little confused by the buildpack putting binaries into `/bin` when my local build puts it in `/` by default. Is this a go-ism that I'm unfamiliar with? cc @hunterloftis 

Does this cause a conflict in the (highly unlikely) case of the user building a binary named `pre-compile` or `post-compile`?

At the risk of bike-shedding, is there a good argument for `bin/post-compile` over `./post-compile` or `heroku/post-compile`? or is this an arbitrary decision?

Reply:
The go-buildpack puts compiled executables in `/bin` and always has (IIRC). This was patterned off `go install`, which puts binaries into `$GOPATH/bin`, so I'm not sure what you mean by you local build puts it in `/`. Do you mean that when you run `go build` the output is placed in `.` ? If so well, the buildpack doesn't use `go build` because `go build` with multiple packages (`./cmd/...`, `./...`, etc) discards the results.

PS: I also generally suggest people not use `go build` locally because of this inconsistency.

--

It's possible that compilation could put an executable named `post-compile` into `/bin`, but IMO that's a feature. i.e. it allows someone to write their post-compile "script" in go, which the buildpack would compile and then execute. So that is the reason I like having it as `/bin/post-compile` and `/bin/pre-compile` for consistency of location. Also, `/bin` is where python keeps them and we shouldn't have 5 different ways to do things.

Which reminds me.... I really prefer `-` over `_`, but python is using `_`, so we should probaby switch to `_` over `-`. Thoughts?



`pre-compile` on the other hand, since it happens before compilation, can't be written in go.

Reply:
> This was patterned off `go install`, which puts binaries into `$GOPATH/bin`, so I'm not sure what you mean by you local build puts it in `/`. Do you mean that when you run `go build` the output is placed in `.` ?

This is using go modules if that helps.

> PS: I also generally suggest people not use `go build` locally because of this inconsistency.

I believe I was going off of https://github.com/golang/go/wiki/Modules#quick-start which uses `go build`

```
$ go build 
$ ./hello

Hello, world.
```

As a beginner it seems odd that when pushed to Heroku the binary is put in a different location. Is this something that has changed with go modules that we should revisit?

> Also, `/bin` is where python keeps them and we shouldn't have 5 different ways to do things.
> Which reminds me.... I really prefer `-` over `_`, but python is using `_`, so we should probably switch to `_` over `-`. Thoughts?

I think we're going to have different ways of doing this per-language regardless since it should conform to the community's conventions. I wouldn't worry so much about being consistent here.

Plus I believe the Python functionality was undocumented, though I do not know why. Perhaps @CaseyFaist does.

> It's possible that compilation could put an executable named `post-compile` into `/bin`, but IMO that's a feature. i.e. it allows someone to write their post-compile "script" in go, which the buildpack would compile and then execute.

I do like this bit

Reply:
> If so well, the buildpack doesn't use `go build` because `go build` with multiple packages (`./cmd/...`, `./...`, etc) discards the results.

Ah, I think I skipped over this, and need to read more

Reply:
> So that is the reason I like having it as `/bin/post-compile` and `/bin/pre-compile` for consistency of location.

Was talking through this today, and someone brought up that this consistency actually works against us in this case. If for some reason someone were to use both the Go and Python buildpacks (or any other buildpack that used the same location for these hooks), then these hooks would be executed twice.

That's not specifically an issue with this since the path name is different, but having `bin/post-compile` and `bin/post_compile` would be a bit odd. 🤔

Reply:
The behavior of `go build` vs. `go install` doesn't change wrt go modules, or at least it doesn't atm. It may in the future. I suspect `go build` is used in most of the docs because Go is (a) trying to get away from $GOPATH and (b) it's simpler for the simple case.

With that said, I still think that the buildpack will continue to need to use `go install`. I can change the location of where it puts binaries though (`GOBIN=${build}`) but that could possibly break some users as well, so it's not a change I want to `just do`.

Re naming conflicts wrt buildpacks and execution of these scripts happening more than once?

How about `/bin/go-post-compile` and `/bin/go-pre-compile` ?